### PR TITLE
The IWYU linter catches uses of fmt::join

### PR DESCRIPTION
**Pull request by SeanCurtis-TRI**
While our build ecosystem is split between fmtlib 11 and previous versions, it is too easy to forget to include fmt/ranges.h to be compatible in v11. So, rather than getting burned repeatedly, we'll add it to our linter. We can revert it when our various libraries match up.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)

### CODE REVIEW

**Review by SeanCurtis-TRI — COMMENTED** [review]
The simplest way to test this:

1. Checkout the PR.
2. Find any file that uses `fmt::join()`.
3. Remove the `#include <fmt/ranges.h>` included in that file.
4. Run lint on the file.

You should get a warning about including the missing include file. 

+(release notes: none) 
+@rpoyner-tri for feature review, please.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-:-ODNtwxR-G-RWeZDTv7i:b-1gm1e1)* status: LGTM missing from assignee rpoyner-tri(platform), needs at least two assigned reviewers

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, LGTM missing from assignee rpoyner-tri(platform), needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODNvYMs9Bj4-fsTaoEG:b-q9vme5) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
> ```Starlark
>             ":patches/upstream/sre_deprecation.patch",
>             ":patches/test_paths.patch",
>             ":patches/include_for_fmt_join.patch",
> ```

nit At the very least, this should be in the `patches/upstream` directory (since _we (Drake)_ are the upstream).  Or even better, it should be a PR to our styleguide repository, with a git sha bump here.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, LGTM missing from assignee rpoyner-tri(platform), needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODNy3oj9q1fNJhpU0Bk:bp68ozw) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit At the very least, this should be in the `patches/upstream` directory (since _we (Drake)_ are the upstream).  Or even better, it should be a PR to our styleguide repository, with a git sha bump here.
</blockquote></details>

Oh yeah.  D'oh!

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, LGTM missing from assignee rpoyner-tri(platform), needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODNy6pJ5k5zK1nOHVOC:b-59tmqi) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Oh yeah.  D'oh!
</blockquote></details>

The primary reason to prefer it here is that when Drake updates its dependencies, Drake can remove the patch and the styleguide will be none the wiser.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, LGTM missing from assignee rpoyner-tri(platform), needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODNyQhUBHzsx3-kg1nK:brzoii) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
> when Drake updates its dependencies

I'm not sure what this means?

Do you mean when we stop using `fmt` at some point?  Or that fmt evolves and the mapping from function-to-header-file will vary depending on the versions of fmt we aim to support?

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm: pending patch discussion

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-:-ODO6UdC2AOJJ7wn8QOy:b-i74iso)* status: 1 unresolved discussion, needs at least two assigned reviewers

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODPGNJz97Ptj08E7v0z:bd3nyqt) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> when Drake updates its dependencies

I'm not sure what this means?

Do you mean when we stop using `fmt` at some point?  Or that fmt evolves and the mapping from function-to-header-file will vary depending on the versions of fmt we aim to support?
</blockquote></details>

What I had in mind specifically is documented at the top of the patch.

tl;dr When we're in a version where the compiler barks about the missing header on all Drake-supported platform, the linting will become redundant. The idea is that it's a transitory thing.

I don't know which will happen first: fmtlib 11 everywhere or std::format. But an event will come that renders this patch pointless.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODRhWUK3KtYX8OIIWx6:b-b30sgr) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

What I had in mind specifically is documented at the top of the patch.

tl;dr When we're in a version where the compiler barks about the missing header on all Drake-supported platform, the linting will become redundant. The idea is that it's a transitory thing.

I don't know which will happen first: fmtlib 11 everywhere or std::format. But an event will come that renders this patch pointless.
</blockquote></details>

I don't buy "pointless" -- or at least, if it's pointless then then the entire build/IWYU cpplint category is pointless:

Imagine the code uses`std::set` but forgot to `#include <set>`.  The cpplint would complain.  However, the code might actually still compile OK, because the `<set>` was indirectly included elsewhere in a way that happened to meet the meed.  But later, when that other code evolves to stop meeting the need, the code that was using `std::set` without following IWYU would suddenly fail to build.

The same holds for `fmt::join` and `fmt/ranges.h` -- even once all of our platforms are using fmt 11, we still might have cases where IWYU was not followed and we are subject to that same trap of a ticking bomb when other faraway code evolves.

If we only cared about pre-merge CI catching it, we already have that scheduled for next week -- once we refresh our mac images, pre-merge CI will have fmt11 so will flag this problem.  This cpplint enhancement will still be relevant because it catches the problem even when the compiler can't.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODSbtgx86bm8MZEJSYj:bpoqv26) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't buy "pointless" -- or at least, if it's pointless then then the entire build/IWYU cpplint category is pointless:

Imagine the code uses`std::set` but forgot to `#include <set>`.  The cpplint would complain.  However, the code might actually still compile OK, because the `<set>` was indirectly included elsewhere in a way that happened to meet the meed.  But later, when that other code evolves to stop meeting the need, the code that was using `std::set` without following IWYU would suddenly fail to build.

The same holds for `fmt::join` and `fmt/ranges.h` -- even once all of our platforms are using fmt 11, we still might have cases where IWYU was not followed and we are subject to that same trap of a ticking bomb when other faraway code evolves.

If we only cared about pre-merge CI catching it, we already have that scheduled for next week -- once we refresh our mac images, pre-merge CI will have fmt11 so will flag this problem.  This cpplint enhancement will still be relevant because it catches the problem even when the compiler can't.
</blockquote></details>

Fair enough. "Pointless" went too far.

You've provided some new data points. I hadn't realized we were so close to being able to catch this problem pre-merge CI. To be perfectly frank, that was the only problem I was solving.

With the new context, the *new* question is: Given that our IWYU enforcement is limited to being an interesting heuristic, do we want to extend it this way? The scenario you provide is true for a multitude of header files. Is there a reason to explicitly include `fmt::join`/`fmt/ranges.h` to the heuristic by itself? Would it make sense to account for all of `fmt::` at the same time? Or none of it?

If not, I'm content to kill this PR and wait for CI changes to solve my original problem.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: 1 unresolved discussion, needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODmeWCK5HLKBjXuYClL:b7gnvgl) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
> ... do we want to extend it this way?

This is your initiative.  I'm happy with whichever outcome you want to choose.

BTW the current schedule for new CI images (that will fail pre-merge) is to roll out this Thursday.

> The scenario you provide is true for a multitude of header files. Is there a reason [for which heuristics we should be checking] ...

The cpplint IWYU checker is most useful for includes that are widely used, and therefore frequently churn in header files.  The baseline (upstream) is only to cover stdlib headers, because anything else would not be portable -- those are the only headers that are relevant for everyone.  I defer to you to decide whether or not `fmt/ranges.h` is worthwhile to add.

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/22276)* status: needs at least two assigned reviewers

___
*[`tools/workspace/styleguide/repository.bzl` line 15 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/22276#-ODNvYMs9Bj4-fsTaoEF:-ODmpIZi3UqiIr7b70oA:ba952zr) ([raw file](https://github.com/RobotLocomotion/drake/blob/01c3f39bd7bacb31fde6a1c7f5d127287dd206e4/tools/workspace/styleguide/repository.bzl#L15)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> ... do we want to extend it this way?

This is your initiative.  I'm happy with whichever outcome you want to choose.

BTW the current schedule for new CI images (that will fail pre-merge) is to roll out this Thursday.

> The scenario you provide is true for a multitude of header files. Is there a reason [for which heuristics we should be checking] ...

The cpplint IWYU checker is most useful for includes that are widely used, and therefore frequently churn in header files.  The baseline (upstream) is only to cover stdlib headers, because anything else would not be portable -- those are the only headers that are relevant for everyone.  I defer to you to decide whether or not `fmt/ranges.h` is worthwhile to add.
</blockquote></details>

I'll close this PR. If the new CI images doesn't catch it (for whatever arcane and mysterious reason), I'll roll it back out. But if it's purpose is satisfied by required changes, there's no point in adding optional changes on top of it.

**Comment by SeanCurtis-TRI** [comment]
As per discussion -- failure to include `fmt/ranges.h` will be caught in CI very soon.
